### ISSUE TITLE
Speed up CI tests

### ISSUE THREAD

**Issue by egpbos** [spokenlanguage/platalea#18]
The test suite as set up in #17 takes over 40 minutes. The major bottlenecks are:

- pip_seq, 18 minutes
- mtl_asr, ~6 min.
- pip_ind, ~5 min.
- mtl_st, ~3:30 min.
- conda install dependencies (mainly due to download of pytorch and deps), ~3:15
- basic & asr, ~2 min.

For the experiments, would it be an idea if we run them with different hyperparameters so they train faster? If so, what parameters should we target?

If this is not enough, there are also a few possibilities within GitHub Actions to speed the whole test suite up:

1. Making the workflow parallel by using multiple jobs and dependencies between them (see commit e46e81f for a not yet working example of this). To do this, we must persist all files, including the whole environment of dependencies, between jobs by using artifacts. Since the PyTorch dependencies alone are over 1 GB, this may cost too much time transferring data back and forth between jobs, but maybe not, we could check.
2. Making a custom VM image with dependencies like PyTorch preinstalled would save a lot of time in setup. Combining this with option 1 would be ideal. Not sure if it's possible though, have to investigate.
3. [Caching](https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows) may be a better solution for storing the environment between parallel jobs. Artifacts can then be used to save coverage information to forward to the final job which calculates the total coverage.

**Comment by bhigy** [comment]
What is the status of this? Has there been some progress since June?

**Comment by egpbos** [comment]
I have modified the `get_default_config` functions used for `pip_seq`, so they take an extra `hidden_size_factor` parameter. With this, the default 1024 can be reduced to say 4. This reduces runtime on my laptop from 16 minutes to 45 seconds, so that seems worth the effort.

The bulk of this is the part under "Extracting ASR/SLT transcriptions", in function `extract_trn`, which by itself takes about 32 seconds (most of the runtime). Most of the runtime there, in turn, is caused by `net.transcribe`. Most of *that* time is caused by `beam_search`. Setting `use_beam_search=False` indeed reduces runtime of that part as well, giving a final total runtime of only 10 seconds.

@bhigy how crucial is beam search to that experiment? Can we somehow change hyperparameters in such a way that it takes less time? Can we switch beam search off for this experiment and somehow test the beam search function separately?

**Comment by egpbos** [comment]
A similar parameter for the `mtl_asr` experiment reduces its runtime (on my laptop) from 4m15s to 9 seconds.

**Comment by egpbos** [comment]
`pip_ind` reduced to 15 seconds. Again mostly constrained by `extract_trn`, but not much to gain here, I guess.

**Comment by egpbos** [comment]
In `mtl_st` it reduces runtime to 5 seconds. Also, a mini bug fixed: in the TextEncoder config, a literal value of 1024 was used instead of the `hidden_size` parameter, which caused mismatching shapes when running with a different `hidden_size`.

**Comment by egpbos** [comment]
`basic_default` reduced to 7 seconds. `asr` reduced to 5 seconds.

**Comment by egpbos** [comment]
Actually, @bhigy, should we just add a general configuration parameter `hidden_size` to `config.py`? I added them to all the experiments seperately for testing now.

**Comment by egpbos** [comment]
Ok, I just went ahead and added this, let me know if you think it should be done differently.

The test suite now takes 9 minutes, of which 6,5 minutes for conda. I'm going to try doing it all with pip now, let's see if that helps even further.

**Comment by egpbos** [comment]
Indeed, this helps a lot, runtime is now down to ~ 4 minutes.

The slowest experiment is still `pip_seq`, with almost 1,5 minutes.

As mentioned above, this could be shortened by deactivating beam_search. Do you think it makes sense to do that @bhigy? Or is the test then useless?

In terms of coverage, I think it is not an issue, because beam search is also used in the `pip_ind` experiment. Or does that experiment cover a different conditional branch of the function?

**Comment by egpbos** [comment]
I'm trying to see whether it decreases coverage in d0aa4fd ([branch skip_beam_search](https://github.com/spokenlanguage/platalea/tree/skip_beam_search)).

**Comment by egpbos** [comment]
Ah, fun, it actually increases coverage :) That must be because the beam search in the `pip_ind` run already covered the same part and turning it off in `pip_seq` actually covers a different part. Also, runtime of `pip_seq` goes down to 6 seconds like this.

**Comment by egpbos** [comment]
We are now down to 2-3 minutes. The main remaining bottleneck is now installation via pip, which took a minute [in the last (2 minute total) run](https://github.com/spokenlanguage/platalea/runs/1690922418?check_suite_focus=true). I don't see any way to speed this up further and also making the tests run in parallel will not make a big difference. So done!
